### PR TITLE
:sparkles: (device) [LIVE-13852]: implement delete app data use case

### PR DESCRIPTION
**Pull request by valpinkman**
### ✅ Checklist

- [x] `npx changeset` was attached.
- [x] **Covered by automatic tests.** 
- [x] **Impact of the changes:** 
  - add new delete app data use case in live-common/device

### 📝 Description

Added a new usecase for the deletion of local app data stored on the device storage (llm/lld)

### ❓ Context

- **JIRA or GitHub link**: [LIVE-13852] 

---

### 🧐 Checklist for the PR Reviewers

- **The code aligns with the requirements** described in the linked JIRA or GitHub issue.
- **The PR description clearly documents the changes** made and explains any technical trade-offs or design decisions.
- **There are no undocumented trade-offs**, technical debt, or maintainability issues.
- **The PR has been tested** thoroughly, and any potential edge cases have been considered and handled.
- **Any new dependencies** have been justified and documented.
- **Performance** considerations have been taken into account. (changes have been profiled or benchmarked if necessary)

[LIVE-13852]: https://ledgerhq.atlassian.net/browse/LIVE-13852?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: libs/ledger-live-common/src/device/use-cases/appDataBackup/types.ts [hunk 1]
```diff
@@ -142,3 +142,31 @@ export type RestoreAppDataEvent =
...
+      type: DeleteAppDataEventType.AppDataDeleted;
+    }
+  | {
+      type: DeleteAppDataEventType.NoAppDataToDelete;
+    };
+
+/**
+ * An error that occurs during the delete process (local data), the error message should be descriptive when thrown.
+ */
+export const DeleteAppDataError = createCustomErrorClass("DeleteAppDataError");
```
Comment (ofreyssinet-ledger) [inline libs/ledger-live-common/src/device/use-cases/appDataBackup/types.ts:172]: [SHOULD] Add translation keys for LLM & LLD for this error (even if there will be more specific error messages)

File: libs/ledger-live-common/src/device/use-cases/appDataBackup/restoreAppData.ts [hunk 2]
```diff
@@ -50,6 +51,8 @@ export function restoreAppData(
           // Commit the restore process, last step
           await restoreAppStorageCommit(transport);
 
+          // NOTE: DELETE DATA
```
Comment (ofreyssinet-ledger) [inline libs/ledger-live-common/src/device/use-cases/appDataBackup/restoreAppData.ts:54]: [ASK] Will be done in another PR?

File: libs/ledger-live-common/src/device/use-cases/appDataBackup/deleteAppDataUseCase.ts [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * @returns Observable<DeleteAppDataEvent>
+ */
+export function deleteAppDataUseCase(
+  appName: AppName,
+  deviceModelId: DeviceModelId,
+  storageProvider: StorageProvider<AppStorageType>,
+  deleteAppDataFn: () => Observable<DeleteAppDataEvent>,
+) {
+  const obs: Observable<DeleteAppDataEvent> = from(deleteAppDataFn());
+  return obs;
```
Comment (ofreyssinet-ledger) [inline libs/ledger-live-common/src/device/use-cases/appDataBackup/deleteAppDataUseCase.ts:21]: [COULD] Remove unused parameters, and do this directly:
```
return deleteAppDataFn()
```

Comment (ofreyssinet-ledger) [reply libs/ledger-live-common/src/device/use-cases/appDataBackup/deleteAppDataUseCase.ts:21]: OR it should be like that:
```
export function deleteAppDataUseCase(
  appName: AppName,
  deviceModelId: DeviceModelId,
  storageProvider: StorageProvider<AppStorageType>,
  deleteAppDataFn: (appName, deviceModelId, storageProvider) => Observable<DeleteAppDataEvent>,
) {
  return deleteAppDataFn(appName, deviceModelId, storageProvider
}
```

Comment (ofreyssinet-ledger) [reply libs/ledger-live-common/src/device/use-cases/appDataBackup/deleteAppDataUseCase.ts:21]: I don't know actually just keep it like that maybe 🤷‍♂️

File: libs/ledger-live-common/src/device/use-cases/appDataBackup/deleteAppData.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        }),
+        switchMap(async item => {
+          if (item) {
+            try {
+              await storageProvider.removeItem(`${deviceModelId}-${appName}`);
+              subscriber.next({ type: DeleteAppDataEventType.AppDataDeleted });
+              subscriber.complete();
+            } catch (e: unknown) {
+              const message = e instanceof Error ? e.message : "Error deleting app data";
+              throwError(() => new DeleteAppDataError(message));
```
Comment (ofreyssinet-ledger) [inline libs/ledger-live-common/src/device/use-cases/appDataBackup/deleteAppData.ts:37]: [SHOULD] this error is not propagated to the returned observable

Comment (jiyuzhuang) [reply libs/ledger-live-common/src/device/use-cases/appDataBackup/deleteAppData.ts:37]: [SHOULD] You can just throw an error instead of using the operator `throwError` according to [the official doc](https://rxjs.dev/api/index/function/throwError#unnecessary-usage).

File: libs/ledger-live-common/src/device/use-cases/appDataBackup/deleteAppDataUseCase.test.ts [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    it("should emit the correct events when the app data is deleted", async () => {
+      deleteAppDataFn.mockReturnValue(
+        concat(
+          of({
+            type: DeleteAppDataEventType.AppDataDeleteStarted,
+          }),
+          of({
+            type: DeleteAppDataEventType.AppDataDeleted,
+          }),
+        ),
```
Comment (ofreyssinet-ledger) [inline libs/ledger-live-common/src/device/use-cases/appDataBackup/deleteAppDataUseCase.test.ts:46]: [COULD] simpler way to do that with:
```ts
of(
  {type: DeleteAppDataEventType.AppDataDeleteStarted},
  {type: DeleteAppDataEventType.AppDataDeleted},
)
```

or
```ts
from([
  {type: DeleteAppDataEventType.AppDataDeleteStarted},
  {type: DeleteAppDataEventType.AppDataDeleted},
])

File: libs/ledger-live-common/src/device/use-cases/appDataBackup/deleteAppData.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+} from "./types";
+
+export function deleteAppData(
+  appName: AppName,
+  deviceModelId: DeviceModelId,
+  storageProvider: StorageProvider<AppStorageType>,
+): Observable<DeleteAppDataEvent> {
+  const obs = new Observable<DeleteAppDataEvent>(subscriber => {
+    subscriber.next({ type: DeleteAppDataEventType.AppDataDeleteStarted });
+    const sub = from(storageProvider.getItem(`${deviceModelId}-${appName}`))
```
Comment (ofreyssinet-ledger) [inline libs/ledger-live-common/src/device/use-cases/appDataBackup/deleteAppData.ts:19]: [SHOULD] Have 1 function (unique source of truth) `getAppDataStorageId` (or similar naming) to get the string `${deviceModelId}-${appName}`, reused throughout backup/delete/restore.

File: libs/ledger-live-common/src/device/use-cases/appDataBackup/deleteAppDataUseCase.ts (hunk 3 above)
Comment (jiyuzhuang) [reply libs/ledger-live-common/src/device/use-cases/appDataBackup/deleteAppDataUseCase.ts:21]: I think that @ofreyssinet-ledger 's proposition is correct, the objectif of this layer is to add the storage specification. The result of the execution of `deleteAppDataFn` is already an observable of type `DeleteAppDataEvent` which should be exposed directly to the components (user interface logic).

**Review by ofreyssinet-ledger — COMMENTED** [review]
lgtm once the change we've discussed are implemented 👍  (error handling & testing)

**Review by jiyuzhuang — APPROVED** [review]

**Review by ofreyssinet-ledger — DISMISSED** [review]

**Review by ofreyssinet-ledger — APPROVED** [review]
